From 2acf168e38884b3a03e3a03a4fbd4ba12447e122 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Thu, 30 Mar 2017 20:14:18 +0800
Subject: [PATCH 1698/5242] MLK-14582: ASoC: imx-pcm-rpmsg: fix audio noise
 issue with pulseaudio

commit  3ecfc540707345daab968f3dc277a5cd4b2775a7 from
https://source.codeaurora.org/external/imx/linux-imx.git

The pulse audio will set a wrong buffer size which is not the integral
multiple of period size, it will cause the DMA can't work correctly in
M4 side.
The reason is that we always need to add a constraint for
SNDRV_PCM_HW_PARAM_PERIODS, which make it integer.

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
(cherry picked from commit 54a10a6c2130a69aca4c1923dac3a15137911146)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-pcm-rpmsg.c |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sound/soc/fsl/imx-pcm-rpmsg.c b/sound/soc/fsl/imx-pcm-rpmsg.c
index 242fa98..7670bcd 100644
--- a/sound/soc/fsl/imx-pcm-rpmsg.c
+++ b/sound/soc/fsl/imx-pcm-rpmsg.c
@@ -142,6 +142,11 @@ static int imx_rpmsg_pcm_open(struct snd_pcm_substream *substream)
 
 	substream->runtime->private_data = prtd;
 
+	ret = snd_pcm_hw_constraint_integer(substream->runtime,
+					    SNDRV_PCM_HW_PARAM_PERIODS);
+	if (ret < 0)
+		return ret;
+
 	return ret;
 }
 
-- 
1.7.9.5

